core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -253,6 +254,39 @@ public void to(MutationComposer composer)
     }
   }
 
+  /**
+   * Returns the read set. This is only for internal use. It doesn't copy the values for performance
+   * reasons.
+   *
+   * @return the read set. Don't modify the value.
+   */
+  @SuppressFBWarnings(value = {"EI_EXPOSE_REP"})
+  public Map<Key, Optional<TransactionResult>> getReadSet() {
+    return readSet;
```

Comment:
How about wrapping it with `new HashMap()` or `ImmutableMap.copyOf()`?

Reply:
The snapshot hook mechanism introduced in this PR could impact the performance and memory consumption of applications that use ScalarDB. So, I think all the operations in a snapshot hook should be fast and low-memory-footprint as much as possible. As described in the method comment, it would be great if a snapshot hook can read those write set and read set without any coping entries.

I also thought the following iteration method, but it's too usage-specific.
```java
  public void iterateWriteSet(BiConsumer<Put, Optional<TransactionResult>> consumer) {
    for (Map.Entry<Key, Put> entry : writeSet.entrySet()) {
      Optional<TransactionResult> result = Optional.empty();
      if (readSet.containsKey(entry.getKey())) {
        result = readSet.get(entry.getKey());
      }
      consumer.accept(entry.getValue(), result);
    }
  }
```

I think using `java.util.Collections#unmodifiableMap()` is a reasonable option since it doesn't have its own table but delegates read operations to the original map. The performance impact should be negligible. What do you think?

Reply:
I see. Sorry, I missed the method comment. That sounds reasonable.

> I think using `java.util.Collections#unmodifiableMap()` is a reasonable option since it doesn't have its own table but delegates read operations to the original map. The performance impact should be negligible. What do you think?

Actually, I’m not sure about the differences in performance between `Collections#unmodifiableMap()`, `new HashMap()`, and `ImmutableMap.copyOf()`, so let's keep it as is for now.

By the way, we already have getter methods for `writeSet` and `deleteSet`, and they copy `writeSet` and `deleteSet`:
https://github.com/scalar-labs/scalardb/blob/037913934c5e7dca44262c280b48136300fdeb92/core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java#L157-L163

For the same reason, we probably don’t want to copy them. So could you please modify these methods as well and add similar method comments, even though it’s not directly related to this PR?

Reply:
> so let's keep it as is for now.

👍 

As for `getPutsInWriteSet()` and `getDeletesInDeleteSet()`, they wrap the maps' values with ArrayList constructor, which also internally copies received values. So, I'll change the return type to `Collection<Put>` and `Collection<Delete>`, respectively.

Reply:
Addressed them in https://github.com/scalar-labs/scalardb/pull/2271/commits/71b40a4168c24dbfdbd8eb4ef7852b071cbc2395

Reply:
I think it might be better to do a simple benchmarking if we care about the performance.
`ImmutableMap.copyOf` actually tries to avoid copying if it is safe as described in the doc, so it might not cause much overhead.
https://guava.dev/releases/23.0/api/docs/com/google/common/collect/ImmutableMap.html#copyOf-java.util.Map-

Reply:
Yeah, thanks for the heads-up. I micro-benchmarked yesterday with the following code. Sorry, I should've mentioned this benchmark.

```java
package org.example;

import com.google.common.collect.ImmutableMap;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class IteratorTest {
  private static final Map<String, Integer> MAP = new HashMap<>();
  static {
    MAP.put("one", 1);
    MAP.put("two", 2);
    MAP.put("three", 3);
    MAP.put("four", 4);
    MAP.put("five", 5);
    MAP.put("six", 6);
    MAP.put("seven", 7);
    MAP.put("eight", 8);
    MAP.put("nine", 9);
    MAP.put("ten", 10);
  }

  private static void testDirect() {
    Map<String, Integer> map = MAP;
    for (Map.Entry<String, Integer> entry : map.entrySet()) {
      entry.getKey();
      entry.getValue();
    }
  }

  private static void testUnmod() {
    Map<String, Integer> map = Collections.unmodifiableMap(MAP);
    for (Map.Entry<String, Integer> entry : map.entrySet()) {
      entry.getKey();
      entry.getValue();
    }
  }

  private static void testNewMap() {
    Map<String, Integer> map = new HashMap<>(MAP);
    for (Map.Entry<String, Integer> entry : map.entrySet()) {
      entry.getKey();
      entry.getValue();
    }
  }

  private static void testImmutable() {
    Map<String, Integer> map = ImmutableMap.copyOf(MAP);
    for (Map.Entry<String, Integer> entry : map.entrySet()) {
      entry.getKey();
      entry.getValue();
    }
  }

  public static void main(String[] args) {
    int n = 1000000;

    for (int i = 0; i < n; i++) { testDirect(); }

    {
      long start = System.currentTimeMillis();
      for (int i = 0; i < n; i++) {
        testDirect();
      }
      System.out.println("Direct: duration=" + (System.currentTimeMillis() - start));
    }

    for (int i = 0; i < n; i++) { testUnmod(); }
    {
      long start = System.currentTimeMillis();
      for (int i = 0; i < n; i++) {
        testUnmod();
      }
      System.out.println("UnmodiMap: duration=" + (System.currentTimeMillis() - start));
    }

    for (int i = 0; i < n; i++) { testNewMap(); }
    {
      long start = System.currentTimeMillis();
      for (int i = 0; i < n; i++) {
        testNewMap();
      }
      System.out.println("NewMap: duration=" + (System.currentTimeMillis() - start));
    }

    for (int i = 0; i < n; i++) { testImmutable(); }
    {
      long start = System.currentTimeMillis();
      for (int i = 0; i < n; i++) {
        testImmutable();
      }
      System.out.println("Immutable: duration=" + (System.currentTimeMillis() - start));
    }
  }
}
```

The result was:
```
> Task :IteratorTest.main()
Direct: duration=20
UnmodiMap: duration=21
NewMap: duration=89
Immutable: duration=130
```

As for the Javadoc, I also looked into the implementation. It doesn't copy only when the original map is `ImmutableMap` instance. In the case of `Snapshot`. the original maps are `HashMap` and I think they'll be copied.